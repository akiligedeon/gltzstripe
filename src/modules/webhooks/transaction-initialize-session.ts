import {
  getEnvironmentFromKey,
  getStripeExternalUrlForIntentId,
  initializeStripePaymentIntent,
  stripePaymentIntentToTransactionResult,
  transactionSessionInitializeEventToStripeCreate,
} from "../stripe/stripe-api";
import { getSaleorAmountFromStripeAmount } from "../stripe/currencies";
import { getWebhookPaymentAppConfigurator } from "../payment-app-configuration/payment-app-configuration-factory";
import { paymentAppFullyConfiguredEntrySchema } from "../payment-app-configuration/config-entry";
import { obfuscateConfig, obfuscateValue } from "../app-configuration/utils";
import { getConfigurationForChannel } from "../payment-app-configuration/payment-app-configuration";
import { type TransactionInitializeSessionResponse } from "@/schemas/TransactionInitializeSession/TransactionInitializeSessionResponse.mjs";
import { type TransactionInitializeSessionEventFragment } from "generated/graphql";
import { invariant } from "@/lib/invariant";
import { createLogger } from "@/lib/logger";

export const TransactionInitializeSessionWebhookHandler = async (
  event: TransactionInitializeSessionEventFragment,
  saleorApiUrl: string,
): Promise<TransactionInitializeSessionResponse> => {
  const logger = createLogger(
    { saleorApiUrl },
    { msgPrefix: "[TransactionInitializeSessionWebhookHandler] " },
  );
  logger.debug(
    {
      transaction: event.transaction,
      action: event.action,
      sourceObject: {
        id: event.sourceObject.id,
        channel: event.sourceObject.channel,
        __typename: event.sourceObject.__typename,
      },
      merchantReference: event.merchantReference,
    },
    "Received event",
  );

  const app = event.recipient;
  invariant(app, "Missing event.recipient!");

  const { privateMetadata } = app;
  const configurator = getWebhookPaymentAppConfigurator({ privateMetadata }, saleorApiUrl);
  const appConfig = await configurator.getConfig();

  const stripeConfig = paymentAppFullyConfiguredEntrySchema.parse(
    getConfigurationForChannel(appConfig, event.sourceObject.channel.id),
  );

  logger.info({}, "Processing Transaction Initialize request");

  const paymentIntentCreateParams = transactionSessionInitializeEventToStripeCreate(event);
  logger.debug({
    paymentIntentCreateParams: obfuscateConfig(paymentIntentCreateParams),
    environment: getEnvironmentFromKey(stripeConfig.publishableKey),
  });

  const stripePaymentIntent = await initializeStripePaymentIntent({
    paymentIntentCreateParams,
    secretKey: stripeConfig.secretKey,
  });

  const data = {
    paymentIntent: { client_secret: stripePaymentIntent.client_secret },
    publishableKey: stripeConfig.publishableKey,
  };
  logger.debug(
    {
      paymentIntent: {
        client_secret: data.paymentIntent.client_secret
          ? obfuscateValue(data.paymentIntent.client_secret)
          : "",
      },
      publishableKey: obfuscateValue(data.publishableKey),
    },
    "Transaction Initialize response",
  );

  const result = stripePaymentIntentToTransactionResult(
    event.action.actionType,
    stripePaymentIntent,
  );
  logger.debug(result, "Stripe -> Transaction result");

  const transactionInitializeSessionResponse: TransactionInitializeSessionResponse = {
    data,
    pspReference: stripePaymentIntent.id,
    result,
    amount: stripePaymentIntent.amount
      ? getSaleorAmountFromStripeAmount({
          amount: stripePaymentIntent.amount,
          currency: stripePaymentIntent.currency,
        })
      : 0,
    time: stripePaymentIntent.created
      ? new Date(stripePaymentIntent.created * 1000).toISOString()
      : undefined,
    message: stripePaymentIntent.cancellation_reason || stripePaymentIntent.description || "",
    externalUrl: stripePaymentIntent.id
      ? getStripeExternalUrlForIntentId(stripePaymentIntent.id)
      : undefined,
  };

  return transactionInitializeSessionResponse;
};
