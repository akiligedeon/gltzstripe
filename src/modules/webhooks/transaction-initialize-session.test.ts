import { describe, it, expect } from "vitest";

import { TransactionInitializeSessionWebhookHandler } from "./transaction-initialize-session";
import {
  createMockTransactionInitializeSessionEvent,
  createMockTransactionInitializeSessionSourceObjectCheckout,
  createMockTransactionInitializeSessionSourceObjectOrder,
} from "./__tests__/utils";
import { setupRecording } from "@/__tests__/polly";
import { testEnv } from "@/__tests__/test-env.mjs";

import { TransactionFlowStrategyEnum } from "generated/graphql";

describe(`TransactionInitializeSessionWebhookHandler`, () => {
  setupRecording({
    matchRequestsBy: {
      headers: {
        exclude: [
          "date",
          "idempotency-key",
          "original-request",
          "request-id",
          "content-length",
          "x-stripe-client-user-agent",
        ],
      },
      method: false,
      body: false,
      order: false,
      url: false,
    },
  });

  describe.each([
    {
      name: "Checkout",
      getSourceObject: createMockTransactionInitializeSessionSourceObjectCheckout,
    },
    { name: "Order", getSourceObject: createMockTransactionInitializeSessionSourceObjectOrder },
  ])(`$name`, ({ getSourceObject }) => {
    it.each([
      {
        title: `should work no action required`,
        data: {},
        result: "AUTHORIZATION_REQUESTED",
        amount: 99.99 + 123.0,
        actionType: TransactionFlowStrategyEnum.Authorization,
      },
      {
        title: `should work no action required`,
        data: {},
        result: "CHARGE_REQUESTED",
        amount: 99.99 + 123.0,
        actionType: TransactionFlowStrategyEnum.Charge,
      },
    ])(`$title`, async ({ title, data, result, amount, actionType }) => {
      const event = await createMockTransactionInitializeSessionEvent({
        data,
        sourceObject: getSourceObject(),
        action: {
          actionType,
        },
      });
      const initializeResult = await TransactionInitializeSessionWebhookHandler(
        event,
        testEnv.TEST_SALEOR_API_URL,
      );
      expect(initializeResult.data).toEqual(expect.any(Object));
      expect(initializeResult.result).toEqual(result);
      expect(initializeResult.amount).toEqual(amount);
      expect(initializeResult.data).toMatchSnapshot(title);
    });
  });
});
