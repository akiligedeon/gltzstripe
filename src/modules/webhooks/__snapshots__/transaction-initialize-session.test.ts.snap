// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionInitializeSessionWebhookHandler > 'Checkout' > 'should work no action required' > should work no action required 1`] = `
{
  "paymentIntent": {
    "client_secret": "pi_3NIs7aEosEcNBN5m1qys2Hp3_secret_tgNpZAiBVzkU1Pfq3dt7Foka4",
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionInitializeSessionWebhookHandler > 'Checkout' > 'should work no action required' > should work no action required 2`] = `
{
  "paymentIntent": {
    "client_secret": "pi_3NIs7aEosEcNBN5m1qys2Hp3_secret_tgNpZAiBVzkU1Pfq3dt7Foka4",
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionInitializeSessionWebhookHandler > 'Order' > 'should work no action required' > should work no action required 1`] = `
{
  "paymentIntent": {
    "client_secret": "pi_3NIs7aEosEcNBN5m1CsbLROi_secret_qmblwAclIVorRBPBiLdv5KiiO",
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionInitializeSessionWebhookHandler > 'Order' > 'should work no action required' > should work no action required 2`] = `
{
  "paymentIntent": {
    "client_secret": "pi_3NIs7aEosEcNBN5m1CsbLROi_secret_qmblwAclIVorRBPBiLdv5KiiO",
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;
